src/pages/resultsView/bookmark/BookmarkModal.tsx

```diff
@@ -2,71 +2,124 @@ import * as React from "react";
-import Loader from "../../../shared/components/loadingIndicator/LoadingIndicator";
+import Loader, {default as LoadingIndicator} from "../../../shared/components/loadingIndicator/LoadingIndicator";
 import {Modal} from 'react-bootstrap';
 import ExtendedRouterStore from "../../../shared/lib/ExtendedRouterStore";
 import {remoteData} from "../../../shared/api/remoteData";
 import getBrowserWindow from "../../../shared/lib/getBrowserWindow";
 import URL, {QueryParams} from 'url';
+import {observable} from "mobx";
+import {ShareUrls} from "../querySummary/ShareUI";
+var Clipboard = require('clipboard');
 
 @observer
-export class BookmarkModal extends React.Component<{ onHide: () => void, routingStore: ExtendedRouterStore }, {}> {
+export class BookmarkModal extends React.Component<{ onHide: () => void, urlPromise:Promise<any> }, {}> {
 
     constructor(){
         super();
     }
 
-    shortenedUrl = remoteData(async () => {
-
-        const win = getBrowserWindow();
-
-        let sessionUrl;
-        // if we have a session use current url
-        if (this.props.routingStore._session) {
-            sessionUrl = getBrowserWindow().location.href;
-        } else {
-            const resp = await this.props.routingStore.saveRemoteSession(this.props.routingStore.query);
-            sessionUrl = URL.format({
-                hostname: win.location.hostname,
-                pathname: win.location.pathname,
-                protocol: win.location.protocol,
-                query: {
-                    session_id: resp.id
-                }
-            });
-        }
+    @observable
+    urlData:ShareUrls;
+
+    clipboards:any[] = [];
+
+    private clipboardInitialized = false;
+
+    componentDidMount(){
+        // this is an $.ajax promise, not a real promise
+        this.props.urlPromise.then(
+            (data)=>{
+                this.urlData = data;
+            },
+            ()=>{
+
+            }
+        );
+
+    }
 
-        const BITLY_ACCESS_TOKEN = "7fdf316a43437f8a6291b99963b371d0229d4b2e";
+    componentDidUpdate(){
+
+        // this is so messed up but is necessary because of terrible copy to clipboard library issues when used inside modal
+        setTimeout(()=>{
+            if (!this.clipboardInitialized && this.urlData && this.bitlyButton && this.sessionButton) {
+                this.clipboardInitialized = true;
+
+                this.clipboards.push(new Clipboard(this.sessionButton, {
+                    text: function() {
+                        return this.urlData.sessionUrl || this.urlData.fullUrl;
+                    }.bind(this),
+                    container: this.container
+                }));
+
+                this.clipboards.push(new Clipboard(this.bitlyButton, {
+                    text: function() {
+                        return this.urlData.bitlyUrl;
+                    }.bind(this),
+                    container: this.container
+                }));
+            }
+        },1000)
```

Comment:
I don't understand the one second timeout.. if it can be 0, then I think it should be that. If it can't because you're trying to give time for something to be available, maybe instead make a setInterval that checks whether some condition is satisfied and then executes this?